langchain/src/chains/combine_documents/refine.ts

```diff
@@ -0,0 +1,144 @@
+    strings generated by the chain, in the order that they were generated.
+ *
+ * @example
+ * ```ts
+ * import { ChatOpenAI } from "@langchain/openai";
+ * import { ChatPromptTemplate } from "@langchain_core/prompts/chat";
+ * 
+ * const llm = new ChatOpenAI();
+ * const initialPrompt = ChatPromptTemplate.fromMessages([
+ *  ["system", "Summarize this information:"],
+ *  ["user", {context}"]
+ * ])
+ * const refinePrompt = ChatPromptTemplate.fromMessages([
+ *  ["system", `You are summarizing a long document one page at a time.
+      You have summarized part of the document. Given the next page, update your
+      summary. Respond with only the updated summary and no other text.
+      Here is your working summary:\n\n{output}.`],
+ *  ["user", "Here is the next page:\n\n{context}"]
+ * ])
+ * 
+ * const chain = await createRefineDocumentsChain(llm, initialPrompt, refinePrompt);
+ * const output = await chain.invoke({ context: documents, question: "..." });
+ */
+export async function createRefineDocumentsChain(
+  llm: LanguageModelLike,
+  initialPrompt: BasePromptTemplate,
+  refinePrompt: BasePromptTemplate,
+  documentPrompt: BasePromptTemplate = DEFAULT_DOCUMENT_PROMPT
+) {
+  if (!initialPrompt.inputVariables.includes(DOCUMENTS_KEY)) {
+    throw new Error(`Initial prompt must include ${DOCUMENTS_KEY} variable`);
+  }
+  if (!refinePrompt.inputVariables.includes(DOCUMENTS_KEY)) {
+    throw new Error(`Refine prompt must include ${DOCUMENTS_KEY} variable`);
+  }
+  if (!refinePrompt.inputVariables.includes(OUTPUT_KEY)) {
+    throw new Error(`Refine prompt must include ${OUTPUT_KEY} variable`);
+  }
+  // work on first document in the list, creating first intermediate output
+  const initialStep = RunnablePassthrough.assign({
+    [OUTPUT_KEY]: RunnableSequence.from(
+      [
+        formatNextDoc.bind(null, documentPrompt),
+        initialPrompt,
+        llm,
+        new StringOutputParser(),
+      ],
+      "initial_step"
+    ),
+  });
+  // work on next document in the list, creating next intermediate output
+  const refineStep = RunnablePassthrough.assign({
+    [OUTPUT_KEY]: RunnableSequence.from(
+      [
+        formatNextDoc.bind(null, documentPrompt),
+        refinePrompt,
+        llm,
+        new StringOutputParser(),
+      ],
+      "refine_step"
+    ),
+    [INTERMEDIATE_STEPS_KEY]: (inputs: IntermediateInputs) => [
+      ...inputs[INTERMEDIATE_STEPS_KEY],
+      inputs[OUTPUT_KEY],
+    ],
+  });
+  return RunnableSequence.from(
+    [
+      // first document
+      initialStep,
+      // remaining documents
+      (inputs: Record<string, Document[]>) => {
+        const docs = inputs[DOCUMENTS_KEY];
+        if (docs.length <= 1) {
+          return new RunnablePassthrough();
+        } else if (docs.length === 2) {
+          return refineStep;
+        } else {
+          return RunnableSequence.from(
+            docs.slice(1).map(() => refineStep) as unknown as [
```

Comment:
Such a hard cast is necessary? Could we type `inputs[DOCUMENTS_KEY]` to `Runnable[]` instead?

Reply:
No the issue is the array length. There's no way to make typescript understand an array map has at least 2 elements